emukit/bayesian_optimization/local_penalization_calculator.py

```diff
@@ -0,0 +1,93 @@
+
+N_SAMPLES = 500  # Number of samples to use when estimating Lipschitz constant
+MAX_ITER = 200  # Maximum number of iterations for optimizer when estimating Lipschitz constant
+
+
+class LocalPenalizationPointCalculator(CandidatePointCalculator):
+    """
+    Candidate point calculator that computes a batch using local penalization from:
+
+    `Batch Bayesian Optimization via Local Penalization. Javier González, Zhenwen Dai, Philipp Hennig, Neil D. Lawrence
+    <https://arxiv.org/abs/1505.08052>`_
+    """
+
+    def __init__(self, acquisition: Acquisition, acquisition_optimizer: AcquisitionOptimizer, model: IDifferentiable,
+                 parameter_space: ParameterSpace, batch_size: int):
+        """
+        :param acquisition: Base acquisition function to use without any penalization applied, this acquisition should
+                            output positive values only.
+        :param acquisition_optimizer: AcquisitionOptimizer object to optimize the penalized acquisition
+        :param model: Model object, used to compute the parameters of the local penalization
+        :param parameter_space: Parameter space describing input domain
+        :param batch_size: Number of points to collect in each batch
+        """
+        self.acquisition = acquisition
+        self.acquisition_optimizer = acquisition_optimizer
+        self.batch_size = batch_size
+        self.model = model
+        self.parameter_space = parameter_space
+
+    def compute_next_points(self, loop_state: LoopState) -> np.ndarray:
+        """
+        Computes a batch of points using local penalization.
+
+        :param loop_state: Object containing the current state of the loop
+        """
+
+        # Initialize local penalization acquisition
+        local_penalization_acquisition = LocalPenalization(self.model)
+
+        # Everything done in log space so addition here is same as multiplying acquisition with local penalization
+        # function.
+        acquisition = self.acquisition + local_penalization_acquisition
+
+        x_batch = []
+        for i in range(self.batch_size):
+            # Collect point
+            x_next, _ = self.acquisition_optimizer.optimize(acquisition)
+            x_batch.append(x_next)
+
+            # Update local penalization acquisition with x_next
+            f_min = np.min(self.model.Y)
+            lipschitz_constant = _estimate_lipschitz_constant(self.parameter_space, self.model)
+            local_penalization_acquisition.update_batches(np.concatenate(x_batch, axis=0), lipschitz_constant, f_min)
+        return np.concatenate(x_batch, axis=0)
+
+
+def _estimate_lipschitz_constant(space: ParameterSpace, model: IDifferentiable):
+    """
+    Estimate the lipschitz constant of the function by max norm of gradient currently in the model. Find this max
+    gradient norm using an optimizer.
+    """
+    def negative_gradient_norm(x):
+        x = np.atleast_2d(x)
```

Comment:
Can we delete this line? If not, where are we not sure about the dimentionality?

Reply:
Removed, I'll use a lambda function to make sure scipy sees a function with a 1d input still